### DISCUSSION TITLE
What is the method to configure fallback font for kitty?

### DISCUSSION THREAD

**Discussion by mendaxcat** [kovidgoyal/kitty#4007] [Q&A] [answered]
I noticed that some chinese characters are not displayed correctly in kitty.
`kitty --debug-font-fallback` outputs
```
U+8fbe Face(family=Noto Sans CJK JP, style=Regular, ps_name=NotoSansCJKjp-Regular, path=/usr/share/fonts/opentype/noto/NotoSansCJK-Regular.ttc, index=0, is_scalable=True, has_color=False, ascender=1160, descender=-288, height=1448, underline_position=-150, underline_thickness=50, strikethrough_position=325, strikethrough_thickness=50) (new face)
U+95e8 Face(family=Noto Sans CJK JP, style=Regular, ps_name=NotoSansCJKjp-Regular, path=/usr/share/fonts/opentype/noto/NotoSansCJK-Regular.ttc, index=0, is_scalable=True, has_color=False, ascender=1160, descender=-288, height=1448, underline_position=-150, underline_thickness=50, strikethrough_position=325, strikethrough_thickness=50)
```
So kitty finds "Noto Sans CJK JP" as its fallback font, is there any way to set "Noto Sans CJK SC" as the fallback font for kitty?

Actually I have already set the fontconfig like
```
<alias>
	<family>sans-serif</family>
	<prefer>
		<family>Noto Sans</family>
		<family>Noto Sans CJK SC</family>
		<family>Noto Sans CJK TC</family>
		<family>Noto Sans CJK JP</family>
		<family>Noto Sans CJK KR</family>
		<family>Noto Color Emoji</family>
	</prefer>
</alias>
```
It works well for other programs.

![image](https://user-images.githubusercontent.com/90375986/132640492-baa878db-143b-410d-836a-99a4ee041025.png)
the left part of the picture is kitty, the right part is the expected result.

I am using kitty 0.23.1 on debian 11.

**Comment by kovidgoyal** [comment, marked as answer]
https://sw.kovidgoyal.net/kitty/conf/#opt-kitty.symbol_map

**Comment by toastal** [comment]
I don't think the current answer is acceptable. Not all Unicode ranges are known for what symbols are "missing". I was expecting the terminal would fallback to fontconfig or allow multiple values like CSS.

NOTE: I'm not looking for Nerd Fonts hacks or "programming ligatures" or foreign languages specifically, these _are_ answered by the above.

**Reply by kovidgoyal** [reply]
kitty does use the platform's blessed library for finding fonts, which on Linux is fontconfig and macOS is CoreText. I am just not interested in teaching people how to use fontconfig. For instance the OP here is defining fallbacks for the sans-serif family, which is irrelevant to a terminal since terminals look for monospaced fonts. Read create_fallback_face in fontconfig.c for details.

**Comment by yingmanwumen** [comment]
> I noticed that some chinese characters are not displayed correctly in kitty. `kitty --debug-font-fallback` outputs
> 
> ```
> U+8fbe Face(family=Noto Sans CJK JP, style=Regular, ps_name=NotoSansCJKjp-Regular, path=/usr/share/fonts/opentype/noto/NotoSansCJK-Regular.ttc, index=0, is_scalable=True, has_color=False, ascender=1160, descender=-288, height=1448, underline_position=-150, underline_thickness=50, strikethrough_position=325, strikethrough_thickness=50) (new face)
> U+95e8 Face(family=Noto Sans CJK JP, style=Regular, ps_name=NotoSansCJKjp-Regular, path=/usr/share/fonts/opentype/noto/NotoSansCJK-Regular.ttc, index=0, is_scalable=True, has_color=False, ascender=1160, descender=-288, height=1448, underline_position=-150, underline_thickness=50, strikethrough_position=325, strikethrough_thickness=50)
> ```
> 
> So kitty finds "Noto Sans CJK JP" as its fallback font, is there any way to set "Noto Sans CJK SC" as the fallback font for kitty?
> 
> Actually I have already set the fontconfig like
> 
> ```
> <alias>
> 	<family>sans-serif</family>
> 	<prefer>
> 		<family>Noto Sans</family>
> 		<family>Noto Sans CJK SC</family>
> 		<family>Noto Sans CJK TC</family>
> 		<family>Noto Sans CJK JP</family>
> 		<family>Noto Sans CJK KR</family>
> 		<family>Noto Color Emoji</family>
> 	</prefer>
> </alias>
> ```
> 
> It works well for other programs.
> 
> ![image](https://user-images.githubusercontent.com/90375986/132640492-baa878db-143b-410d-836a-99a4ee041025.png) the left part of the picture is kitty, the right part is the expected result.
> 
> I am using kitty 0.23.1 on debian 11.

I've the same problem with you. Try this:

```
 <match target="pattern">
   <test name="prgname" compare="eq">
     <string>kitty</string>
   </test>
   <edit name="family" mode="prepend" binding="strong">
     <string>FiraCode Nerd Font</string>
   <family>Noto Sans CJK SC</family>
   <family>Noto Sans CJK TC</family>
   <family>Noto Sans CJK JP</family>
   <family>Noto Sans CJK KR</family>
   </edit>
 </match>
 <alias>
  <family>FiraCode Nerd Font</family>
  <prefer>
   <family>FiraCode Nerd Font</family>
   <family>Noto Sans CJK SC</family>
   <family>Noto Sans CJK TC</family>
   <family>Noto Sans CJK JP</family>
   <family>Noto Sans CJK KR</family>
   <family>Noto Color Emoji</family>
   <family/>
  </prefer>
 </alias>
```

**Reply by chris4540** [reply]
The font configuration works well.

Here is my config for Japanese monospace char

`cat ~/.config/fontconfig/fonts.conf`

```xml
<?xml version="1.0"?>
<!DOCTYPE fontconfig SYSTEM "urn:fontconfig:fonts.dtd">
<fontconfig>
    <match target="pattern">
        <test name="prgname" compare="eq">
            <string>kitty</string>
        </test>
        <edit mode="prepend" name="family">
            <string>Utatane</string>
        </edit>
    </match>
</fontconfig>
```
